ssl/quic/quic_port.c

```diff
@@ -35,28 +35,39 @@ static void port_rx_pre(QUIC_PORT *port);
  * @brief Represents a validation token for secure connection handling.
  *
  * This struct is used to store information related to a validation token,
- * including the token buffer, original connection ID, and an integrity tag
- * for secure validation of QUIC connections.
+ * including an integrity tag for secure validation of QUIC connections.
  *
- * @var validation_token::token_buf
- * A character array holding the token data. The size of this array is
- * based on the length of the string "openssltoken" minus one for the null
- * terminator.
+ * @var validation_token::is_retry
+ * True iff this validation token is for a token sent in a RETRY packet.
+ * Otherwise, this token is from a NEW_TOKEN_packet. Iff this value is true,
+ * then ODCID and RSCID are set.
  *
- * @var validation_token::token_odcid
+ * @var validation_token::timestamp
+ * Time that the validation token was minted.
+ *
+ * @var validation_token::remote_addr_len
+ * Length of the following character array.
+ *
+ * @var validation_token::remote_addr
+ * A character array holding the raw address of the client requesting the
+ * connection.
+ *
+ * @var validation_token::odcid
  * An original connection ID (`QUIC_CONN_ID`) used to identify the QUIC
  * connection. This ID helps associate the token with a specific connection.
  *
- * @var validation_token::integrity_tag
- * A character array for the integrity tag, with a length defined by
- * `QUIC_RETRY_INTEGRITY_TAG_LEN`. This tag is used to verify the integrity
- * of the token during the connection process.
+ * @var validation_token::rscid
+ * The original retry source connection ID (`QUIC_CONN_ID`).
  */
-struct validation_token {
-    char token_buf[sizeof("openssltoken") - 1];
-    QUIC_CONN_ID token_odcid;
-    char integrity_tag[QUIC_RETRY_INTEGRITY_TAG_LEN];
-};
+typedef struct validation_token {
+    int is_retry;
+    OSSL_TIME timestamp;
+    size_t remote_addr_len;
+    unsigned char *remote_addr;
```

Comment:
I think remote_addr should be part of the token (defined in-place). It will make AEAD() validation easier in my opinion. you can re-arrange the members:
```
int is_retry;
OSSL_TIME timestamp;
QUIC_CONN_ID odcid;
QUIC_CONN_ID rscid;
size_t remote_addr_len;
char addr;
```
then one can allocate token like `token = OPENSSL_malloc(sieof (QUIC_VALIDATION_TOKEN) + addr_len - 1)`
To write/read address, you do memcpy(&token->addr, remote_addr, addr_len);

Reply:
I can do that. However, when parsing/reading the QUIC_CONN_ID's, if I take @nhorman's suggestion of not using the full QUIC_CONN_ID, I won't be able to do `token = OPENSSL_malloc(sieof (QUIC_VALIDATION_TOKEN) + addr_len - 1)`.

Reply:
How does the whole address fit in a single `char addr`? Also, this might need some workarounds if I do @nhorman's suggestion of using only part of a QUIC_CONN_ID

Reply:
I think the last field is meant to be char *addr.  the size of the token would then be:
```
sizeof(int) + sizeof(OSSL_TIME) + odcid.id_len + sizeof(odcid.id_len) + rscid.id_len + sizeof(rscid.id_len) + remote_addr_len
```
Then , when you marshall the data, the format looks like this:
```
 ┌────────────┬──────────┬──────┬──────────────┬─────┬───────────────┬──────┬────────────────────┐  
 │is_retry    │ timestamp│      │              │     │               │ alen │ remote_addr        │  
 │            │          │ idlen│   odcid(var) │idlen│ rscid(var)    │      │                    │  
 └────────────┴──────────┼──────┴──────────────┼─────┴───────────────┴──────┼────────────────────┘  
     
```

you don't actually need the char *addr pointer when you transmit the token, you just copy the BIO_ADDR into the above marshalled buffer.  You only need to use the char on receipt of the token back, so that when you decode it, you can point it to the offset of remote_addr into the buffer;